### PR TITLE
Feature/fpe

### PR DESCRIPTION
**Pull request by HariharanIT**
# Related Issue
- FPE Feature

# Propossed changes/Fix
- Change 1
- Change 2
- ...

# Additional Info 
- Any additional info or context for the reviewer

# Checklist
- [ ] Tested on local/dev branch for all major platforms (Android, IOS, Desktop, Web).
- [ ] No commented out code
- [ ] Is any third party library, service used
- [ ] Tests
- [ ] If this change requires updates outside of the code, like updates in core, react-ui-kit, RTM/RTC configure.

# Dependent PRs 
- Mention dependent PR links.

# Screenshots

Original                |          Updated
:---------------------:  | :-----------------------:
**original screenshot** |  **updated screenshot**

### CODE REVIEW

**Review by nitte93 — CHANGES_REQUESTED** [review]

File: template/babel.config.js [hunk 1]
```diff
@@ -10,8 +10,24 @@
...
-  plugins: [['transform-define', configVars]],
+  plugins: [
+    ['transform-define', configVars],
+    [
+      require.resolve('babel-plugin-module-resolver'),
+      {
+        root: ["./"],
+        alias: {
+          "fpe-api/install": "./fpe-api/install.ts",
+          "fpe-api": "./fpe-api/index.ts",
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/babel.config.js:26]: test-fpe and fpe-api both require each other module. that's why for install method we are having separate alias name. we will discuss more detail over the call

File: template/src/App.tsx [hunk 2]
```diff
@@ -39,9 +43,17 @@ if (Platform.OS === 'ios') {
...
-
+  const {
+    videoCall,
+    join,
+    create
+  } = useFpe(data => data?.components ? data.components : {} as ComponentsInterface);
+  const CustomRoutes = useFpe(data => data?.customRoutes);
+  const CreateCmp = cmpTypeGuard(Create,create);
+  const AppRoot = useFpe(data => data?.appRoot)
+  const RootWrapper = AppRoot && typeof AppRoot === 'function' ? AppRoot : React.Fragment;
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/App.tsx]: Added getTypeGurad method

File: template/src/App.tsx [hunk 3]
```diff
@@ -54,8 +66,33 @@ const App: React.FC = () => {
             <Router>
               <SessionProvider>
                 <ColorConfigure>
+                  <DimensionProvider>
+                  <LanguageProvider>
+                  <ErrorProvider>
+                  <Error />
                   <Navigation />
                   <Switch>
+                    {CustomRoutes?.map((e:CustomRoutesInterface, i: number) => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/App.tsx]: done

File: template/_gitignore [hunk 4]
```diff
@@ -71,4 +71,7 @@ dist/
 #Electron cache
 .electron/
 .webpack/
-out/
\ No newline at end of file
+out/
+
+#FPE plugin folder TODO:hari rename folder later
+template/test-fpe/*
```
Comment (nitte93) [inline template/_gitignore:77]: NOTE: Remove the TODOs and move it to a separate file with the pending items.

File: template/fpe-api/context.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+export const useRtcContext = createHook(RtcContext);
+export const useMinUidContext = createHook(MinUidContext);
+export const useMaxUidContext = createHook(MaxUidContext);
+export const usePropsContext = createHook(PropsContext);
+export const useLocalContext = createHook(LocalContext);
+export const useDeviceContext = createHook(DeviceContext);
+export const useStorageContext = createHook(StorageContext);
+
+export {LocalUserContext} from '../agora-rn-uikit';
```
Comment (nitte93) [inline template/fpe-api/context.ts:23]: why this Provider needs to be exported seperately but not the others
TODO: Remove

File: template/fpe-api/typeDefinition.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  i18n?: i18nInterface[];
+  /**
+   * Life cycle events
+   */
+  lifecycle?: {
+    useBeforeJoin?: CustomHookType;
+    useBeforeCreate?: CustomHookType;
+  };
+  customUserContext?: {
```
Comment (nitte93) [inline template/fpe-api/typeDefinition.ts:116]: Please add a comment why this is needed ? is it a custom context that user want to use ? what's the difference between appRoot and this customUserContext ?
Can there be multiple customUserContexts ? like useUserContext ? 
TODO: Move away from object

File: template/fpe-api/useFpe.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+export interface FpeProviderInterface {
+  children: React.ReactNode;
+  value: FpeApiInterface;
+}
+
+const FpeProvider = (props: FpeProviderInterface) => {
+  return (
+    <FpeContext.Provider value={props.value}>
+      {true ? props.children : <></>}
```
Comment (nitte93) [inline template/fpe-api/useFpe.tsx:28]: again, why a true statement ? remove it from other places as well
```suggestion
      {props.children}
```

File: template/src/App.tsx [hunk 8]
```diff
@@ -27,9 +27,24 @@ import {ImageBackground, Platform, SafeAreaView, StatusBar} from 'react-native';
...
+import {
+  useFpe,
+  CustomRoutesInterface,
+  CUSTOM_ROUTES_PREFIX,
+  FpeProvider,
+  fpeConfig,
+} from 'fpe-api';
+import {LanguageProvider} from './language/useLanguage';
 
 if (Platform.OS === 'ios') {
```
Comment (nitte93) [inline template/src/App.tsx:49]: update with exported utils here

File: template/src/App.tsx [hunk 9]
```diff
@@ -39,58 +54,129 @@ if (Platform.OS === 'ios') {
...
-                    </Route>
-                  </Switch>
-                </ColorConfigure>
-              </SessionProvider>
-            </Router>
-          </GraphQLProvider>
-        </StorageProvider>
-      </SafeAreaView>
-    </ImageBackground>
+    <RootWrapper>
```
Comment (nitte93) [inline template/src/App.tsx:70]: Take this out as a separate appConfigure provider components and use it by passing chindren props,
<RootWrapper>
      <ImageBackground
        source={{uri: $config.BG}}
        style={{flex: 1}}
        resizeMode={'cover'}>
        <SafeAreaView style={{flex: 1}}>
          <StatusBar hidden={true} />
          <Toast ref={(ref) => Toast.setRef(ref)} config={ToastConfig} />
          <StorageProvider>
            <GraphQLProvider>
              <Router>
                <SessionProvider>
                  <ColorConfigure>
                    <DimensionProvider>
                      <LanguageProvider>
                        <ErrorProvider>
                          <Error />
                          <Navigation />
                          <Switch>

File: template/src/components/Chat.tsx [hunk 10]
```diff
@@ -141,7 +150,11 @@ const Chat = (props: any) => {
           <View>
             <View style={style.chatInputContainer}>
               <View style={[style.chatInputLineSeparator, {opacity: 0.3}]} />
-              <ChatInput privateActive={privateActive} />
+              {ChatInputFpe ? (
```
Comment (nitte93) [inline template/src/components/Chat.tsx:153]: use  getTypeGuard(FPE,  ) ?

File: template/src/components/Controls.native.tsx [hunk 11]
```diff
@@ -29,9 +22,13 @@ import {
 } from '../../agora-rn-uikit';
 import Recording from '../subComponents/Recording';
 import LiveStreamControls from './livestream/views/LiveStreamControls';
+import {useVideoCall} from 'fpe-api';
```
Comment (nitte93) [inline template/src/components/Controls.native.tsx:25]: Minor:  if we want to use it like this, then do we wanna call it fpe-api ? from what i understand fpe-api are fpe facing and not to be used inside the app, we should not import from fpe-api but from the actual implementation file.

File: template/src/components/Navbar.tsx [hunk 12]
```diff
@@ -127,7 +128,7 @@ const Navbar = (props: any) => {
       <View
         style={[
           style.roomNameContainer,
-          Platform.OS === 'web' && !mobileAndTabletCheck()
+          Platform.OS === 'web' && !isMobileOrTablet()
```
Comment (nitte93) [inline template/src/components/Navbar.tsx:131]: while we are at it. let's also update here with the utils

File: template/src/components/Navbar.tsx [hunk 13]
```diff
@@ -249,15 +250,35 @@ const Navbar = (props: any) => {
...
-                );
-              }}
-              name={layout ? 'pinnedLayoutIcon' : 'gridLayoutIcon'}
-            />
+            {layouts[layout].iconName ? (
+              <BtnTemplate
+                style={style.btnHolder}
+                onPress={() => {
+                  setLayout((l: Layout) => {
+                    if (l < layouts?.length - 1) {
```
Comment (nitte93) [inline template/src/components/Navbar.tsx:258]: Layout dropdown implementation

File: template/src/components/Share.tsx [hunk 14]
```diff
@@ -30,76 +30,101 @@ import icons from '../assets/icons';
...
-Attendee Meeting ID: ${urlView}
-Host Meeting ID: ${urlHost}`)
-      : (stringToCopy += `Meeting - ${roomTitle}
-Meeting URL: ${urlHost}`);
-
-    pstn
-      ? (stringToCopy += `PSTN Number: ${pstn.number}
-PSTN Pin: ${pstn.dtmf}`)
-      : '';
+    Toast.show({text1: copiedToClipboardText, visibilityTime: 1000});
```
Comment (nitte93) [inline template/src/components/Share.tsx:66]: can you break it down further into separate blocks here

File: template/src/components/useChatUI.tsx [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+*********************************************
+*/
+import {createHook} from 'fpe-implementation';
+import React,{createContext, SetStateAction} from 'react';
+
+export interface privateMsgLastSeenInterface {
+  userId: string | number,
+  lastSeenCount: number
+}
+export interface ChatUIDataInterface {
```
Comment (nitte93) [inline template/src/components/useChatUI.tsx:19]: Exporting this as it is, I'm not sure how helpful this is gonna be for the users, Add it as a pending task to be discussed later after the merge. 
TODO; Discuss

File: template/src/language/default-labels/videoCallScreenLabels.ts [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+        return 'Loading';
+      default:
+        return 'Loading';
+    }
+  },
+  meetingInviteText: ({meetingName, id, url, pstn}) => {
+    let inviteContent = '';
+    if (url) {
+      if (url?.host) {
+        inviteContent += `Meeting - ${meetingName}\nURL for Attendee: ${url?.attendee}\nURL for Host: ${url?.host}`;
```
Comment (nitte93) [inline template/src/language/default-labels/videoCallScreenLabels.ts:151]: break this down into chunks, this is getting messy and hard to read

File: template/src/language/useLanguage.tsx [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+  //then we will update the localstorage value to default value
+  let storedCode = i18nData?.find(
+    (item) => item.locale === store.selectedLanguageCode,
+  )
+    ? store.selectedLanguageCode
+    : undefined;
+
+  const [languageCode, setLanguageCodeLocal] = useState(
+    storedCode ||
+      (i18nData && i18nData.length ? i18nData[0].locale : false) ||
```
Comment (nitte93) [inline template/src/language/useLanguage.tsx:46]: Minor:

```suggestion
      ( Array.isArray(i18nData) && i18nData.length ? i18nData[0].locale : false) ||
```

File: template/src/pages/video-call/CustomUserContextHolder.tsx [hunk 18]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import {useFpe} from 'fpe-api';
+
+const CustomUserContextHolder: React.FC<{children: any}> = (props) => {
+  const useFpeCustomContext = useFpe((config) => config?.customUserContext?.useUserContext);
+  const Component = useFpeCustomContext ? useFpeCustomContext() : null;
```
Comment (nitte93) [inline template/src/pages/video-call/CustomUserContextHolder.tsx:6]: Discuss: useUserContext

File: template/src/pages/video-call/VideoArrayRenderer.tsx [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    typeof config?.components?.videoCall?.renderComponentObject === 'object'
+      ? config?.components?.videoCall?.renderComponentObject
+      : undefined,
+  );
+
+  const minArray = min.map((user, index) => {
+    const MinComponent =
+      FpeRenderComponent &&
+      FpeRenderComponent[user.type] &&
+      isValidElementType(FpeRenderComponent[user.type])
```
Comment (nitte93) [inline template/src/pages/video-call/VideoArrayRenderer.tsx:22]: make it more readable

File: template/src/pages/video-call/VideoArrayRenderer.tsx [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+        ? FpeRenderComponent[user.type]
+        : RenderComponent[user.type]
+        ? RenderComponent[user.type]
+        : RenderComponent['rtc'];
+    return <MinComponent user={user} isMax={false} index={index} />;
+  });
+
+  const maxArray = max.map((user, index) => {
+    const MaxComponent =
+      FpeRenderComponent &&
```
Comment (nitte93) [inline template/src/pages/video-call/VideoArrayRenderer.tsx:32]: the component part is same for both, why not just reuse it ?

File: template/src/pages/video-call/VideoRenderer.tsx [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+          <TextWithTooltip
+            value={
+              user.uid === 'local'
+                ? userList[localUid]
+                  ? userList[localUid].name + ' '
+                  : 'You '
+                : userList[user.uid]
+                ? userList[user.uid].name + ' '
+                : user.uid === 1
+                ? userList[localUid].name + "'s screen "
```
Comment (nitte93) [inline template/src/pages/video-call/VideoRenderer.tsx:83]: why is this not using useString ?

File: template/src/subComponents/ChatContainer.tsx [hunk 22]
```diff
@@ -70,23 +77,39 @@ const ChatContainer = (props: any) => {
...
+                timestamp={message.ts}
                 uid={message.uid}
                 key={message.ts}
               />
             );
           })
         ) : privateMessageStore[selectedUserID] ? (
           privateMessageStore[selectedUserID].map((message: any) => {
-            return (
+            return ChatBubbleFpe ? (
```
Comment (nitte93) [inline template/src/subComponents/ChatContainer.tsx:100]: use cmpTypeGuard here as well

File: template/src/subComponents/CopyJoinInfo.tsx [hunk 23]
```diff
@@ -36,39 +39,52 @@ const SHARE = gql`
...
-        stringToCopy += `\nPSTN Number: ${data.share.pstn.number}\nPSTN Pin: ${data.share.pstn.dtmf}`;
-      }
+      let baseURL =
+        platform === 'web'
+          ? $config.FRONTEND_ENDPOINT || window.location.origin
+          : null;
+      let stringToCopy = meetingInviteText({
+        meetingName: data.share.title,
+        url: baseURL
+          ? data.share.passphrase.host
```
Comment (nitte93) [inline template/src/subComponents/CopyJoinInfo.tsx:62]: Copy ones are a little painful to read everywhere, let's put then and organise them in one.

File: template/src/utils/common.tsx [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+const hasBrandLogo: boolean = !!$config.LOGO;
+
+const shouldAuthenticate: boolean =
+  $config.ENABLE_APPLE_OAUTH ||
+  $config.ENABLE_GOOGLE_OAUTH ||
+  $config.ENABLE_MICROSOFT_OAUTH ||
+  $config.ENABLE_SLACK_OAUTH;
+
+const isWeb = Platform.OS === 'web';
```
Comment (nitte93) [inline template/src/utils/common.tsx:47]: Move isWeb, isAndroid, isIOS to util function and export

File: template/src/utils/isPSTNUser.ts [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+*/
+import {useContext} from 'react';
+import ChatContext from '../../src/components/ChatContext';
+
+/**
+ * This hook will return the function to check whether the current user is a PSTN user or not
+ * @returns function
+ */
+function useIsPSTN() {
+  const {userList} = useContext(ChatContext);
```
Comment (nitte93) [inline template/src/utils/isPSTNUser.ts:20]: why no use use, make sure this consistence of accessing userlist is same everywhere
```suggestion
    const userList = useUserList();;
```

File: template/src/utils/isHostUser.ts [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+ Use without a license or in violation of any license terms and conditions (including use for 
+ any purpose competitive to Agora Lab, Inc.’s business) is strictly prohibited. For more 
+ information visit https://appbuilder.agora.io. 
+*********************************************
+*/
+import {ClientRole} from '../../agora-rn-uikit';
+import {useUserList} from 'fpe-api';
+
+function useIsHost() {
+  const userList = useUserList();
```
Comment (nitte93) [inline template/src/utils/isHostUser.ts:16]: few places you're using useUserList and few places you[re using useContext(ChatContext);
Please useUserList to be consistent.

File: template/src/utils/mobileWebTest.tsx [hunk 27]
```diff
@@ -1,7 +1,7 @@
-const mobileAndTabletCheck = () => {
+const isMobileOrTablet = () => {
     let check = false;
```
Comment (nitte93) [inline template/src/utils/mobileWebTest.tsx:2]: this shouldn't be a .tsx file and also just rename it isMobileOrTablet

File: template/src/utils/useGroupMessages.ts [hunk 28]
```diff
@@ -0,0 +1,24 @@
...
+*/
+import {useContext} from 'react';
+import ChatContext from '../../src/components/ChatContext';
+
+/**
+ * This hook will fetch the group messages
+ * @returns userList
+ */
+function useGroupMessages() {
+  const {messageStore} = useContext(ChatContext);
```
Comment (nitte93) [inline template/src/utils/useGroupMessages.ts:20]: With all use of useContext like this, wrap it within a conditional, do the same for usePrivateMessage as well
```suggestion
  const context = useContext(ChatContext);
  if (context === undefined) {
    throw new Error(`useGroupMessages must be used within a ChatContext`);
  }
  return context;

```

File: template/src/utils/useString.ts [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+import {useFpe} from 'fpe-api';
+import {useLanguage} from '../language/useLanguage';
+import {DEFAULT_I18_DATA} from '../language';
+import {TextDataInterface} from '../language/default-labels/index';
+
+export function usei18nData(
+  selectedLanguageCode: string = DEFAULT_I18_DATA.locale,
+) {
+  const languageData = useFpe((data) => data?.i18n);
+  if (languageData && languageData.length) {
```
Comment (nitte93) [inline template/src/utils/useString.ts:21]: I think you want to make sure if languageData is an array, for that use Array.isArray. 
.length will exist on objects as well,

File: template/src/utils/useString.ts [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+    if (selectedLanguageCode) {
+      let selectedLanguageData = languageData.find(
```
Comment (nitte93) [inline template/src/utils/useString.ts:23]: We'll need to refactor and organise this a bit, 

- if languageData array doesnt exist do early return of default data
- if !selectedData or !selectedLanguageData do early return. 
- else do the rest of the things

File: template/src/utils/useString.ts [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+  const textData = usei18nData(lanCode);
+
+  const getString = (input?: T) => {
+    if (textData[keyName]) {
+      let keyValue = textData[keyName];
+      if (typeof keyValue === 'function') {
+        return keyValue(input);
+      } else if (typeof keyValue === 'string') {
+        return keyValue;
+      } else {
```
Comment (nitte93) [inline template/src/utils/useString.ts:56]: always do early returns, easier to read. 
if !textData[keyName] return ""
if typeOf textData[keyName] === function
if typeof textData[keyName] === string

File: template/src/utils/useUserList.ts [hunk 32]
```diff
@@ -0,0 +1,24 @@
...
+*********************************************
+*/
+import {useContext} from 'react';
+import ChatContext from '../../src/components/ChatContext';
+
+/**
+ * This hook will fetch the user list
+ * @returns userList
+ */
+function useUserList() {
```
Comment (nitte93) [inline template/src/utils/useUserList.ts:19]: Whereever you're trying to access or export a util to useContext, First Always validate it it's used within the provider,
```suggestion
 export function useUserList() {
   const context = useContext(ChatContext);
   if (context === undefined) {
     throw new Error(`useUserList must be used within a Chat Provider`);
   }
   return context;
 }

```

File: template/fpe-api/useFpe.tsx (hunk 7 above)
Comment (HariharanIT) [reply template/fpe-api/useFpe.tsx:28]: done

File: template/src/App.tsx (hunk 8 above)
Comment (HariharanIT) [reply template/src/App.tsx:49]: done

File: template/src/App.tsx (hunk 9 above)
Comment (HariharanIT) [reply template/src/App.tsx:70]: Done

File: template/src/components/Chat.tsx (hunk 10 above)
Comment (HariharanIT) [reply template/src/components/Chat.tsx:153]: Done

File: template/src/components/Controls.native.tsx (hunk 11 above)
Comment (HariharanIT) [reply template/src/components/Controls.native.tsx:25]: done

File: template/src/components/Navbar.tsx (hunk 12 above)
Comment (HariharanIT) [reply template/src/components/Navbar.tsx:132]: Done

File: template/src/subComponents/ChatContainer.tsx (hunk 22 above)
Comment (HariharanIT) [reply template/src/subComponents/ChatContainer.tsx:100]: Done

File: template/src/utils/common.tsx (hunk 24 above)
Comment (HariharanIT) [reply template/src/utils/common.tsx:47]: Done

File: template/src/utils/isPSTNUser.ts (hunk 25 above)
Comment (HariharanIT) [reply template/src/utils/isPSTNUser.ts:20]: Done

File: template/src/utils/mobileWebTest.tsx (hunk 27 above)
Comment (HariharanIT) [reply template/src/utils/mobileWebTest.tsx:2]: Done

File: template/fpe-api/typeDefinition.ts (hunk 6 above)
Comment (HariharanIT) [reply template/fpe-api/typeDefinition.ts]: Done.

File: template/src/language/useLanguage.tsx (hunk 17 above)
Comment (HariharanIT) [reply template/src/language/useLanguage.tsx:46]: done

File: template/src/pages/video-call/VideoArrayRenderer.tsx (hunk 19 above)
Comment (HariharanIT) [reply template/src/pages/video-call/VideoArrayRenderer.tsx:22]: done

File: template/src/pages/video-call/VideoArrayRenderer.tsx (hunk 20 above)
Comment (HariharanIT) [reply template/src/pages/video-call/VideoArrayRenderer.tsx:32]: done

File: template/src/pages/video-call/VideoRenderer.tsx (hunk 21 above)
Comment (HariharanIT) [reply template/src/pages/video-call/VideoRenderer.tsx:83]: Done

File: template/src/utils/useString.ts (hunk 29 above)
Comment (HariharanIT) [reply template/src/utils/useString.ts:21]: done

File: template/src/utils/useString.ts (hunk 31 above)
Comment (HariharanIT) [reply template/src/utils/useString.ts:56]: done

File: template/src/utils/useString.ts (hunk 30 above)
Comment (HariharanIT) [reply template/src/utils/useString.ts:23]: done

File: template/src/components/Share.tsx (hunk 14 above)
Comment (HariharanIT) [reply template/src/components/Share.tsx:67]: Done

File: template/src/utils/useUserList.ts (hunk 32 above)
Comment (HariharanIT) [reply template/src/utils/useUserList.ts:19]: useContext will return the default vaule. so we can't identify whether is wrapped with the provider or not. we will discuss on this one

File: template/src/utils/useGroupMessages.ts (hunk 28 above)
Comment (HariharanIT) [reply template/src/utils/useGroupMessages.ts:20]: useContext will return the default vaule. so we can't identify whether is wrapped with the provider or not. we will discuss on this one

File: template/src/components/Chat.tsx [hunk 33]
```diff
@@ -25,8 +25,13 @@ import ChatInput from '../subComponents/ChatInput';
...
 import ColorContext from './ColorContext';
 import chatContext from './ChatContext';
+import {UserType} from './RTMConfigure';
+import TextWithTooltip from '../subComponents/TextWithTooltip';
+import { useChatUIData, useFpe } from 'fpe-api';
+import { useString } from '../utils/useString';
 
-const Chat = (props: any) => {
+const Chat = () => {
+  const ChatInputFpe = useFpe(data => typeof data.components?.videoCall === 'object' && typeof data.components?.videoCall?.chat === 'object' ? data.components?.videoCall?.chat?.chatInput : undefined)
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/Chat.tsx]: Either type didn't work. we need to try out a different solutions.

File: template/fpe-api/context.ts (hunk 5 above)
Comment (HariharanIT) [reply template/fpe-api/context.ts]: Done

File: template/_gitignore (hunk 4 above)
Comment (HariharanIT) [reply template/_gitignore:77]: Done

File: template/src/components/useChatUI.tsx (hunk 15 above)
Comment (HariharanIT) [reply template/src/components/useChatUI.tsx:19]: Added in todo list

File: template/src/language/default-labels/videoCallScreenLabels.ts (hunk 16 above)
Comment (HariharanIT) [reply template/src/language/default-labels/videoCallScreenLabels.ts:151]: done

File: template/src/pages/video-call/CustomUserContextHolder.tsx (hunk 18 above)
Comment (HariharanIT) [reply template/src/pages/video-call/CustomUserContextHolder.tsx:6]: Added in todo list

File: template/src/components/Navbar.tsx (hunk 13 above)
Comment (HariharanIT) [reply template/src/components/Navbar.tsx:257]: Added into todo list

File: template/src/components/Chat.tsx [hunk 34]
```diff
@@ -141,7 +145,7 @@ const Chat = (props: any) => {
           <View>
             <View style={style.chatInputContainer}>
               <View style={[style.chatInputLineSeparator, {opacity: 0.3}]} />
-              <ChatInput privateActive={privateActive} />
+              {ChatInputFpe ? <ChatInputFpe privateActive={privateActive}/> : <ChatInput privateActive={privateActive} />}
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/Chat.tsx]: done

File: template/src/App.tsx [hunk 35]
```diff
@@ -54,8 +66,33 @@ const App: React.FC = () => {
...
+                            path={CUSTOM_ROUTES_PREFIX + e.path}
+                            exact={e.exact}
+                            key={i}
+                            failureRedirectTo={e.failureRedirectTo ? e.failureRedirectTo : '/'}
+                            {...e.routeProps}                            
+                          >
+                            <e.component {...e.componentProps} />
+                          </PrivateRoute>
+                        );
+                      } else {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/App.tsx]: Updated

**Pull request closed without merging**